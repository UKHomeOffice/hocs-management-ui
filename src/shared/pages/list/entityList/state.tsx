import EntityListItem from '../../../models/entityListItem';

export interface State {
    entities: EntityListItem[];
    entitiesLoaded: boolean;
}
