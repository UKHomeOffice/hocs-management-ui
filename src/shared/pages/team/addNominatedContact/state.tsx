import Item from '../../../models/item';
export interface State {
    inputValue: string;
    selectedContacts: Item[];
    teamName?: string;
    teamUUID: string;
}
