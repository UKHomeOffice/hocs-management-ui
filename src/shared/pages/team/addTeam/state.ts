import Team from '../../../models/team';

export interface State {
    team: Team;
}
