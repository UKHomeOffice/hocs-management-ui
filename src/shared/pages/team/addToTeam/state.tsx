import Item from '../../../models/item';
export interface State {
    inputValue: string;
    selectedUser?: Item | string;
    selectedUsers: Item[];
    teamName?: string;
    users: Item[];
}
