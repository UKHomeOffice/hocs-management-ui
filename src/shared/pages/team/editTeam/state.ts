import Item from '../../../models/item';

export interface State {
    currentDisplayName: string,
    newDisplayName: string
    unit?: Item;
    initialUnit?: Item;
}
