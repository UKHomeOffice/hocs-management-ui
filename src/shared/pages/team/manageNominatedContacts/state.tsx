import Item from '../../../models/item';
export interface State {
    contacts: Item[];
}
