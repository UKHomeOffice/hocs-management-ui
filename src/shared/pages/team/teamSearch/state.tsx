export interface State {
    teamUUID: string;
}
