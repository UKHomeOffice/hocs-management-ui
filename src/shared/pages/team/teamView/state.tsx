import Item from '../../../models/item';

export interface State {
    teamMembers: Item[];
    teamMembersLoaded: boolean;
    teamName?: string;
    unitName?: string;
    active?: boolean;
}
