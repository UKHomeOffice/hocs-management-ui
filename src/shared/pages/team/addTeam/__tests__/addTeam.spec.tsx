import { act, fireEvent, getByText, render, RenderResult, waitFor } from '@testing-library/react';
import { match, MemoryRouter } from 'react-router-dom';
import React from 'react';
import WrappedAddTeam from '../addTeam';
import { createBrowserHistory, History, Location } from 'history';
import * as UnitsService from '../../../../services/unitsService';
import * as useError from '../../../../hooks/useError';
import { State } from '../state';
import {
    GENERAL_ERROR_TITLE, LOAD_UNITS_ERROR_DESCRIPTION,
    TEAM_CREATION_FAILURE_NAME_ALREADY_EXISTS, TEAM_CREATION_FAILURE_UNKNOWN_ERROR,
    VALIDATION_ERROR_TITLE
} from '../../../../models/constants';
import Unit from '../../../../models/unit';
import * as TeamsService from '../../../../services/teamsService';
import * as Validation from '../../../../validation';

let match: match<any>;
let history: History<any>;
let location: Location;
let mockState: State;

const renderComponent = () => render(
    <MemoryRouter>
        <WrappedAddTeam history={history}  location={location} match={match}/>
    </MemoryRouter>
);

const getUnitsSpy = jest.spyOn(UnitsService, 'getUnits');
const getAddTeamSpy = jest.spyOn(TeamsService, 'addTeam');
const useReducerSpy = jest.spyOn(React, 'useReducer');
const reducerDispatch = jest.fn();
const useErrorSpy = jest.spyOn(useError, 'default');
const addFormErrorSpy = jest.fn();
const clearErrorsSpy = jest.fn();
const setMessageSpy = jest.fn();
const validateSpy: jest.SpyInstance = jest.spyOn(Validation, 'validate');

const mockUnits: Unit = {
    displayName: '__someDisplayName1__',
    shortCode: '__someShortCode',
    type: '__someType__',
    value: '__someValue__'
};

beforeEach(() => {
    history = createBrowserHistory();
    location = {
        hash: '',
        key: '',
        pathname: '',
        search: '',
        state: {}
    };
    mockState = {
        team: {
            displayName: '',
            permissions: [
                {
                    accessLevel: 'OWNER',
                    caseTypeCode: ''
                }
            ],
            letterName: '',
            type: '',
            active: true,
            unitUUID: ''
        }
    };

    useReducerSpy.mockImplementation(() => [mockState, reducerDispatch]);
    useErrorSpy.mockImplementation(() => [{}, addFormErrorSpy, clearErrorsSpy, setMessageSpy]);
    reducerDispatch.mockReset();
    addFormErrorSpy.mockReset();
    clearErrorsSpy.mockReset();
    setMessageSpy.mockReset();
});

describe('when the AddTeam component is mounted', () => {
    it('should render with default props', async () => {
        expect.assertions(3);
        let wrapper: RenderResult;

        const expectedInitialState: State = {
            team: {
                displayName: '',
                permissions: [
                    {
                        accessLevel: 'OWNER',
                        caseTypeCode: ''
                    }
                ],
                letterName: '',
                type: '',
                active: true,
                unitUUID: ''
            }
        };

        act(() => {
            wrapper = renderComponent();
        });

        await waitFor(() => {
            expect(mockState).toEqual(expectedInitialState);
            expect(getUnitsSpy).toHaveBeenCalledTimes(1);
            expect(wrapper.container).toMatchSnapshot();
        });
    });

    it('should return an error if the call to retrieve the units fail', async () => {
        expect.assertions(2);
        getUnitsSpy.mockImplementation(() => Promise.reject('error'));

        renderComponent();
        await waitFor(() => {
            expect(setMessageSpy).toBeCalledWith(
                { title: GENERAL_ERROR_TITLE, description: LOAD_UNITS_ERROR_DESCRIPTION }
            );
        });

    });

    it('should return a success message if team creation is successful', async () => {
        expect.assertions(1);
        let wrapper: RenderResult;
        getUnitsSpy.mockImplementation(() => Promise.resolve([mockUnits]));
        getAddTeamSpy.mockImplementationOnce(() => Promise.resolve({ response: { status: 200 } }));
        validateSpy.mockReturnValue(true);

        act(() => {
            wrapper = renderComponent();
        });

        await waitFor(async () => {
            const submitButton = getByText(wrapper.container, 'Add');
            fireEvent.click(submitButton);
        });
        await waitFor(() => {
            expect(history.location.state).toEqual({ successMessage: 'The team was created successfully.' });
        });

    });

    it('should set the correct error message if AddTeam returns a 409 response', async () => {
        expect.assertions(1);
        let wrapper: RenderResult;
        getUnitsSpy.mockImplementation(() => Promise.resolve([mockUnits]));
        getAddTeamSpy.mockImplementationOnce(() => Promise.reject({ response: { status: 409 } }));
        validateSpy.mockReturnValue(true);

        act(() => {
            wrapper = renderComponent();
        });

        await waitFor(async () => {
            const submitButton = getByText(wrapper.container, 'Add');
            fireEvent.click(submitButton);
        });
        await waitFor(() => {
            expect(setMessageSpy).toBeCalledWith(
                { title: VALIDATION_ERROR_TITLE, description: TEAM_CREATION_FAILURE_NAME_ALREADY_EXISTS }
            );
        });
    });

    it('should set the correct error message if AddTeam fails for an unknown reason', async () => {
        expect.assertions(1);
        let wrapper: RenderResult;
        getUnitsSpy.mockImplementation(() => Promise.resolve([mockUnits]));
        getAddTeamSpy.mockImplementationOnce(() => Promise.reject({ response: { status: 500 } }));
        validateSpy.mockReturnValue(true);

        act(() => {
            wrapper = renderComponent();
        });

        await waitFor(async () => {
            const submitButton = getByText(wrapper.container, 'Add');
            fireEvent.click(submitButton);
        });
        await waitFor(() => {
            expect(setMessageSpy).toBeCalledWith(
                { title: GENERAL_ERROR_TITLE, description: TEAM_CREATION_FAILURE_UNKNOWN_ERROR }
            );
        });
    });
});
