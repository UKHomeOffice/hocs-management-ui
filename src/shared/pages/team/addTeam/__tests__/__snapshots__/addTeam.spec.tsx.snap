// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the AddTeam component is mounted should render with default props 1`] = `
<div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-two-thirds-from-desktop"
    >
      <a
        class="govuk-back-link"
        href="/"
      >
        Back
      </a>
      <h1
        class="govuk-heading-xl"
      >
        Create a DCU drafting team
      </h1>
    </div>
  </div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-two-thirds-from-desktop"
    >
      <form
        method="POST"
      >
        <input
          name="_csrf"
          type="hidden"
          value=""
        />
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="teamName"
            id="teamName-label"
          >
            Team name
          </label>
          <input
            class="govuk-input"
            id="teamName"
            name="teamName"
            type="text"
            value=""
          />
        </div>
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="unit-input"
            id="unit-label"
          >
            Unit
          </label>
          <div
            class=" css-b62m3t-container"
            id="unit"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-2-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
              role="log"
            />
            <div
              class="govuk-typeahead__control css-1g8ily9-Control"
            >
              <div
                class="govuk-typeahead__value-container css-pilx6v-ValueContainer"
              >
                <div
                  class="govuk-typeahead__placeholder css-ovmm9z-Placeholder"
                  id="react-select-2-placeholder"
                >
                  Search
                </div>
                <div
                  class="govuk-typeahead__input-container css-qbdosj-Input"
                  data-value=""
                >
                  <input
                    aria-activedescendant=""
                    aria-autocomplete="list"
                    aria-describedby="react-select-2-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    class="govuk-typeahead__input"
                    id="unit-input"
                    role="combobox"
                    spellcheck="false"
                    style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="govuk-typeahead__indicators css-1hb7zxy-IndicatorsContainer"
              >
                <div
                  aria-hidden="true"
                  class="govuk-typeahead__indicator govuk-typeahead__loading-indicator css-pogzpp-loadingIndicator"
                >
                  <span
                    class="css-1xtdfmb-LoadingDot"
                  />
                  <span
                    class="css-zoievk-LoadingDot"
                  />
                  <span
                    class="css-x748d8-LoadingDot"
                  />
                </div>
                <span
                  class="govuk-typeahead__indicator-separator css-62wm9a-IndicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class="govuk-typeahead__indicator govuk-typeahead__dropdown-indicator css-hbt8bd-DropdownIndicator"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="govuk-form-group"
        >
          <button
            class="govuk-button view-team-button"
            type="submit"
          >
            Add
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
